### ISSUE TITLE
Linkerd 2.9 destination component high latency and node topology request failures

### ISSUE THREAD

**Issue by alekseiplotnikov** [linkerd/linkerd2#5684]
### What is the issue?
After upgrading to Linkerd 2.9.2 from 2.7.1 we started seeing random cross deployment communication problems in our  EKS production cluster during peak hours. There are no known issues with services availability, all pods are running. Disabling linkerd proxy injection from a deployment helps to fix the issue. Looking into linkerd-proxy logs we see that many connections are in "fail-fast" mode. After analysing Linkerd Grafana Health dashboard we found that linkerd-destination P95 latency is very high during these times - taking several seconds and often up to 5 seconds. Looking into linkerd-destination logs we see that there are many errors like "Failed to get node topology for node ..." and errors time period match perfectly with incident times.
We have quickly checked the code and it looks like that recent "Topology-Aware Service Routing" feature has introduced increased dependency on Kubernetes API server and in the case of API performance or availability issues it might have an impact on inter-pod communication.
After downgrading Linkerd back to 2.7.1 described problems are gone.

### How can it be reproduced?
Make linkerd-destination node topology requests to Kubernetes API server fail.

### Logs, error output, etc

linkerd-proxy logs:
```
2021-02-02 19:48:19 | [ 252.710896s] WARN ThreadId(01) outbound:accept{peer.addr=10.178.14.208:36092 target.addr=172.20.160.126:33011}: linkerd2_app_core::errors: Failed to proxy request: Service in fail-fast
2021-02-02 19:48:19 | [ 239.922485s] WARN ThreadId(01) outbound:accept{peer.addr=10.178.15.59:53530 target.addr=172.20.81.231:8080}: linkerd2_app_core::errors: Failed to proxy request: Service in fail-fast
2021-02-02 19:48:19 | [ 239.901097s] WARN ThreadId(01) dst: linkerd2_reconnect::service: Service failed error=channel closed
2021-02-02 19:48:18 | [ 251.676478s] WARN ThreadId(01) outbound:accept{peer.addr=10.178.14.208:34412 target.addr=172.20.81.231:8080}: linkerd2_app_core::errors: Failed to proxy request: Service in fail-fast
2021-02-02 19:48:17 | [ 236.835768s] WARN ThreadId(01) outbound:accept{peer.addr=10.178.10.116:35232 target.addr=172.20.160.126:33011}: linkerd2_app_core::errors: Failed to proxy request: Service in fail-fast
2021-02-02 19:48:17 | [ 236.824725s] WARN ThreadId(01) outbound:accept{peer.addr=10.178.10.116:35228 target.addr=172.20.160.126:33011}: linkerd2_app_core::errors: Failed to proxy request: Service in fail-fast
```

linkerd-destination logs
```
2021-02-02 19:48:18 | time="2021-02-02T19:48:18Z" level=error msg="Failed to send address update: rpc error: code = Internal desc = transport: transport: the stream is done or WriteHeader was already called" addr=":8086" component=endpoint-translator remote="127.0.0.1:35458" service="billingservce.default.svc.cluster.local:80"
2021-02-02 19:48:18 | time="2021-02-02T19:48:18Z" level=error msg="Failed to get node topology for node ip-10-178-13-170.eu-west-1.compute.internal: context canceled" addr=":8086" component=endpoint-translator remote="127.0.0.1:35458" service="billlingservice.default.svc.cluster.local:80"
2021-02-02 19:48:17 | time="2021-02-02T19:48:17Z" level=error msg="Failed to send address update: rpc error: code = Internal desc = transport: transport: the stream is done or WriteHeader was already called" addr=":8086" component=endpoint-translator remote="127.0.0.1:35458" service="orderservice.default.svc.cluster.local:8080"
2021-02-02 19:48:17 | time="2021-02-02T19:48:17Z" level=error msg="Failed to get node topology for node ip-10-178-15-76.eu-west-1.compute.internal: context canceled" addr=":8086" component=endpoint-translator remote="127.0.0.1:35458" service="orderservice.default.svc.cluster.local:8080"
2021-02-02 19:48:17 | time="2021-02-02T19:48:17Z" level=error msg="Failed to send address update: rpc error: code = Internal desc = transport: transport: the stream is done or WriteHeader was already called" addr=":8086" component=endpoint-translator remote="127.0.0.1:35458" service="billingservice.default.svc.cluster.local:80"
2021-02-02 19:48:17 | time="2021-02-02T19:48:17Z" level=error msg="Failed to get node topology for node ip-10-178-15-254.eu-west-1.compute.internal: context canceled" addr=":8086" component=endpoint-translator remote="127.0.0.1:35458" service="billingservice.default.svc.cluster.local:80"
```

#### `linkerd check` output

```text
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running
√ can initialize the client
√ can query the control plane API

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist

linkerd-identity
----------------
√ certificate config is valid
√ trust anchors are using supported crypto algorithm
√ trust anchors are within their validity period
√ trust anchors are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
√ issuer cert is valid for at least 60 days
√ issuer cert is issued by the trust anchor

linkerd-webhooks-and-apisvc-tls
-------------------------------
√ tap API server has valid cert
√ tap API server cert is valid for at least 60 days
√ proxy-injector webhook has valid cert
√ proxy-injector cert is valid for at least 60 days
√ sp-validator webhook has valid cert
√ sp-validator cert is valid for at least 60 days

linkerd-api
-----------
√ control plane pods are ready
√ control plane self-check
√ [kubernetes] control plane can talk to Kubernetes
√ [prometheus] control plane can talk to Prometheus
√ tap api service is running

linkerd-version
---------------
√ can determine the latest version
√ cli is up-to-date

control-plane-version
---------------------
√ control plane is up-to-date
√ control plane and cli versions match

linkerd-prometheus
------------------
√ prometheus add-on service account exists
√ prometheus add-on config map exists
√ prometheus pod is running

linkerd-grafana
---------------
√ grafana add-on service account exists
√ grafana add-on config map exists
√ grafana pod is running

Status check results are √
```

### Environment

- Kubernetes Version: 1.18
- Cluster Environment: EKS
- Host OS: Amazon Linux 2
- Linkerd version: 2.9.2

**Comment by mateiidavid** [comment]
Interesting! 

> that there are many errors like "Failed to get node topology for node ..." and errors time period match perfectly with incident times.

I'm not _entirely sure_ what is happening here, but when we fetch a node's topology labels we just make a call to the apiserver. The fact that we're seeing `context cancellation` as an error here might be unrelated to the k8s apiserver itself and related to the context passed in by the server. This is all just at a glance. 

```
2021-02-02 19:48:17 | time="2021-02-02T19:48:17Z" level=error msg="Failed to send address update: rpc error: code = Internal desc = transport: transport: the stream is done or WriteHeader was already called" addr=":8086" component=endpoint-translator remote="127.0.0.1:35458" service="billingservice.default.svc.cluster.local:80"
2021-02-02 19:48:17 | time="2021-02-02T19:48:17Z" level=error msg="Failed to get node topology for node ip-10-178-15-254.eu-west-1.compute.internal: context canceled" addr=":8086" component=endpoint-translator remote="127.0.0.1:35458" service="billingservice.default.svc.cluster.local:80"
```

  Looking at your logs, I can see that the stream is marked as done before you see the topology failure 🤔 


> After downgrading Linkerd back to 2.7.1 described problems are gone.

 The EndpointSlice feature can be turned off so you don't have to downgrade from 2.9.x. In this case, I don't think you're actually using EndpointSlices, when you installed/upgraded did you use the `--enable-endpoint-slices` flag?   
  e.g `linkerd install --enable-endpoint-slices | ...` 

If you have turned it on, then would you mind trying with the flag turned off? I'd like to know if the issue persists.

If the issue here is just with fetching node labels then that can probably be addressed with something like an EndpointSlice check (similar to what we're doing in the watcher).

**Comment by alekseiplotnikov** [comment]
Thanks @mateiidavid for looking into this issue.

> Looking at your logs, I can see that the stream is marked as done before you see the topology failure 🤔

I should have mentioned it, logs are actually in reversed order, as they come from our monitoring system. This is a bit misleading, sorry.

> I don't think you're actually using EndpointSlices, when you installed/upgraded did you use the --enable-endpoint-slices flag?

I think EndpointSlices feature was disabled in our configuration. We installed Linkerd from Helm chart, default value for global.enableEndpointSlices is "false" and we didn't change it.

**Comment by adleong** [comment]
@mateiidavid one thing I notice is that the node lookup uses `k8sClient.CoreV1().Nodes().Get(ctx, srcNode, metav1.GetOptions{})` which I'm guessing just does a direct query to the k8s API.  Since each destination.Get request to the destination service triggers one of these calls, I could imagine at high load we could be overwhelming the k8s API and causing these node lookup requests to become slow and potentially hit timeouts.

I think that maybe switching this to use a NodeInformer would let us take advantage of caching to greatly reduce the number of k8s API requests we make.  What do you think?

**Comment by mateiidavid** [comment]
@adleong that's a very good point! I definitely missed this; (wrongly) assumed that it wouldn't be an issue since we make only one call per pod. I should correlate that with the number of pods running in a cluster the next time.

I think your proposed solution makes lots of sense, will start working on it asap. 👍

**Comment by fpetkovski** [comment]
Hey @mateiidavid, we already applied a patch which completely resolved the problems in our environment. I've submitted a pull request which can either be merged immediately or give you a head start in resolving the issue.

Another thing we noticed is that we get the same 503 errors and a latency spike in destination when we redeploy the control plane. Our investigations show that this is caused by the caches not being warmed up. Adding a 5 second delay to the readiness probe of destination made the problem go away.

**Comment by RagingPuppies** [comment]
> Hey @mateiidavid, we already applied a patch which completely resolved the problems in our environment. I've submitted a pull request which can either be merged immediately or give you a head start in resolving the issue.
> 
> Another thing we noticed is that we get the same 503 errors and a latency spike in destination when we redeploy the control plane. Our investigations show that this is caused by the caches not being warmed up. Adding a 5 second delay to the readiness probe of destination made the problem go away.

is there a way for stupid people like me to apply the fix? such as command to run or whatever?

**Comment by mateiidavid** [comment]
> > Hey @mateiidavid, we already applied a patch which completely resolved the problems in our environment. I've submitted a pull request which can either be merged immediately or give you a head start in resolving the issue.
> > Another thing we noticed is that we get the same 503 errors and a latency spike in destination when we redeploy the control plane. Our investigations show that this is caused by the caches not being warmed up. Adding a 5 second delay to the readiness probe of destination made the problem go away.
> 
> is there a way for stupid people like me to apply the fix? such as command to run or whatever?

Hey @RagingPuppies this should have been fixed in PR #5722 -- think it was shipped in stable 2.9.4.

If you're having similar problems I suggest giving any version (including an edge) > 2.9.4 a try, and if it persists it might be worth opening a new issue.
